### ISSUE TITLE
Lots and lots of string/regex-escape bugs, including charnames

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12022]
Migrated from [rt.perl.org#112084](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=112084) (status was 'resolved')

Searchable as RT112084$

**Comment by p5pRT** [comment]
### From tchrist@perl.com

SUMMARY&#8203;: Errors in \\x\{\.\.\.\} and \\N\{\.\.\.\} are handled poorly\, leading
&nbsp;     to mysteriously inconsistent and incorrect results\.  Also\,
&nbsp;         their are parsing bugs in \\N\{\.\.\.\.\} that permit all kinds of
&nbsp;         stuff that it shouldn't\.
&nbsp;         
This seems wrong&#8203;:

&nbsp;    $ blead \-wle 'print length "\\x\{\}"'
&nbsp;    1
&nbsp;    $ blead \-wle 'print ord "\\x\{\}"'
&nbsp;    0

The answer \*should\* be that that is altogether illegal\.  The stupid
thing isn't even giving me a warning\.  What's that about?

Contrast with this much more reasonable behavior&#8203;:

&nbsp;    $ blead \-wle 'print ord "\\o\{\}"'
&nbsp;    Number with no digits at \-e line 1\, within string
&nbsp;    Execution of \-e aborted due to compilation errors\.

And I cannot see any justification for this&#8203;:

&nbsp;    $ blead \-le 'print $n = \(\) = "8\\x\{00\}8" =~ /\\A\[\\8\]\{3\}\\z/'
&nbsp;    1
&nbsp;    $ blead \-le 'print $n = \(\) = "g\\x\{00\}g" =~ /\\A\[\\xg\]\{3\}\\z/'
&nbsp;    1

Yes\, those will generate optional ignorable warnings if you beg it nicely\,
but they're \*syntax errors\*\.  It doesn't seem right to emit optional and
by\-default\-off warnings on syntax errors\, for goodness' sake\!  The same
thing should happen as with a bad \\o\{\} \-\- it should die\.

This is a poor message\, for several reasons&#8203;:

&nbsp;    % blead \-le 'print "\." =~ /\\N\{\}/ || "FAIL"'
&nbsp;    Unknown charname '' at \-e line 1\.
&nbsp;    Deprecated character in \\N\{\.\.\.\}; marked by \<\-\- HERE  in \\N\{\}\<\-\- HERE  at \-e line 1\.
&nbsp;    FAIL

First\, that looks a great deal like a syntax error to me\.  Why isn't
it marked as such and the whole thing dying?  It's not just that it
couldn't find a character name\.  It's that it wasn't even given one\.
That should be any error\.  

There are other weirdness\.  I understand what's going on\, but it's
a bit surprising\.

&nbsp;    % blead \-le 'print "\." =~ /\\N\{$a\}/ || "FAIL"'
&nbsp;    Unknown charname '$a' at \-e line 1\.
&nbsp;    Deprecated character in \\N\{\.\.\.\}; marked by \<\-\- HERE  in \\N\{$\<\-\- HERE a\} at \-e line 1\.
&nbsp;    FAIL

&nbsp;    % blead \-le '$a = 1; print "\." =~ /\\N\{$a\}/ || "FAIL"'
&nbsp;    Unknown charname '$a' at \-e line 1\.
&nbsp;    Deprecated character in \\N\{\.\.\.\}; marked by \<\-\- HERE  in \\N\{$\<\-\- HERE a\} at \-e line 1\.
&nbsp;    Name "main&#8203;::a" used only once&#8203;: possible typo at \-e line 1\.
&nbsp;    FAIL

Ok\, so variable interpolation doesn't happen in \\N\{brackets\}\.  Is that
actually documented?

And what sort of warning \*is\* that?  How do you control it?

This didn't turn it off&#8203;:

&nbsp;    % blead \-X \-le 'print "\." =~ /\\N\{\}/ || "FAIL"'
&nbsp;    Unknown charname '' at \-e line 1\.
&nbsp;    FAIL

Nor did this&#8203;:

&nbsp;    % blead \-M\-warnings \-le 'print "\." =~ /\\N\{\}/ || "FAIL"'
&nbsp;    Unknown charname '' at \-e line 1\.
&nbsp;    FAIL

And yet this fatalized it&#8203;:

&nbsp;    % blead \-Mwarnings=FATAL\,all \-le 'print "\." =~ /\\N\{\}/ || "FAIL"'
&nbsp;    Unknown charname '' at \-e line 1\.
&nbsp;    Deprecated character in \\N\{\.\.\.\}; marked by \<\-\- HERE  in \\N\{\}\<\-\- HERE  at \-e line 1\.
&nbsp;    Exit 255

And this missed it altogether&#8203;:

&nbsp;    % perl \-Mcharnames=&#8203;:full \-M\-warnings \-le 'print "\." =~ /\\N\{\}/ || "FAIL"'
&nbsp;    Unknown charname '' at \-e line 1

&nbsp;    FAIL

So you can't turn it off in any of the normal ways\.

It also seems to have an extra newline in the output\.

Here diagnostics couldn't catch it&#8203;:

&nbsp;    % blead \-Mcharnames=&#8203;:full \-Mdiagnostics \-le 'print "\." =~ /\\N\{\}/ || "FAIL"'
&nbsp;    Unknown charname '' at \-e line 1

&nbsp;    Deprecated character in \\N\{\.\.\.\}; marked by \<\-\- HERE  in \\N\{\}\<\-\- HERE  at \-e
&nbsp;        line 1 \(\#1\)
&nbsp;    \(D deprecated\) Just about anything is legal for the \.\.\. in \\N\{\.\.\.\}\.
&nbsp;    But starting in 5\.12\, non\-reasonable ones that don't look like names
&nbsp;    are deprecated\.  A reasonable name begins with an alphabetic character
&nbsp;    and continues with any combination of alphanumerics\, dashes\, spaces\,
&nbsp;    parentheses or colons\.
&nbsp;    
&nbsp;    FAIL

And there's all that extra newline business still\.

That error message is misleading\.  It looks like it's saying

&nbsp;    m\{
&nbsp;    \\A \\p\{Alphabetic\} 
&nbsp;    \[\\p\{Alphabetic\}\\p\{Number\}\\p\{Dash\}\\N\{SPACE\}\(\)&#8203;:\] \+
&nbsp;    \\z 
&nbsp;    \}z


But that's incorrect\.  It appears to actually be this&#8203;:

&nbsp;    m\{
&nbsp;    \\A 
&nbsp;    \(?= \\p\{word\} \)            \# must start with a word character\, that isn't digit or \_
&nbsp;    \(?\! \[\\N\{LOW LINE\}\\p\{digit\}\] \)   \# but \*does\* admit other numerics and connector punctuation\!\!
&nbsp;    \[\\p\{word\}\\N\{HYPHEN\-MINUS\}\\N\{SPACE\}\(\)&#8203;:\] \*
&nbsp;    \\z 
&nbsp;    \}x

Which isn't what the message says\.  And that star really should be a 
plus\, as previously discussed\.

Oh wait\.  It turns out that this is the real expressions&#8203;:

&nbsp;    % setenv PERL\_UNICODE=SA

&nbsp;    % perl \-Mcharnames=&#8203;:full \-le 'print qq\(print "\\\\\\N\{X\\N\{EURO SIGN\}\}"\)'
&nbsp;    print "\\N\{X€\}"
&nbsp;    % perl \-Mcharnames=&#8203;:full \-le 'print qq\(print "\\\\\\N\{X\\N\{EURO SIGN\}\}"\)' | blead \-Mutf8
&nbsp;    Unknown charname 'Xâ¬' at \- line 1\.
&nbsp;    ?
&nbsp;    % perl \-le 'print qq\(print "\\\\\\N\{a\\x\{fffd\}\}"\)' | uniquote \-v
&nbsp;    print "\\N\{a\\N\{REPLACEMENT CHARACTER\}\}"
&nbsp;    % perl \-le 'print qq\(print "\\\\\\N\{a\\x\{fffd\}\}"\)' | uniquote \-x
&nbsp;    print "\\N\{a\\x\{FFFD\}\}"
&nbsp;    % perl \-le 'print qq\(print "\\\\\\N\{a\\x\{fffd\}\}"\)' | blead \-Mutf8
&nbsp;    Unknown charname 'aï¿½' at \- line 1\.
&nbsp;    % perl \-le 'print qq\(print "\\\\\\N\{a\\x\{fffd\}\#\#x%\}"\)' | blead \-Mutf8 \-l
&nbsp;    Unknown charname 'aï¿½\#\#x%' at \- line 1\.
&nbsp;    Deprecated character in \\N\{\.\.\.\}; marked by \<\-\- HERE  in \\N\{aï¿½\#\<\-\- HERE \#x%\} at \- line 1\.
&nbsp;    ?
&nbsp;    % perl \-Mcharnames=&#8203;:full \-le 'print qq\(print "\\\\\\N\{X\\N\{EURO SIGN\}\\N\{REPLACEMENT CHARACTER\}\\N\{PESO SIGN\}\\N\{ALIEN MONSTER\}\}"\)' | blead \-Mutf8 \-l
&nbsp;    Unknown charname 'Xâ¬ï¿½â±ð¾' at \- line 1\.
&nbsp;    ?
&nbsp;    % perl \-Mcharnames=&#8203;:full \-le 'print qq\(print "\\\\\\N\{X\\N\{EURO SIGN\}\\N\{REPLACEMENT CHARACTER\}\\N\{PESO SIGN\}\\N\{ALIEN MONSTER\}\.\}"\)' | blead \-Mutf8 \-l
&nbsp;    Unknown charname 'Xâ¬ï¿½â±ð¾\.' at \- line 1\.
&nbsp;    Deprecated character in \\N\{\.\.\.\}; marked by \<\-\- HERE  in \\N\{Xâ¬ï¿½â±ð¾\.\<\-\- HERE \} at \-
&nbsp;    line 1\.
&nbsp;    ?

First\, Perl is still double\-encoding UTF\-8 on stderr\.  Is that ever going
to be fixed?

But most importantly\, this is all messed up\, and in all kinds of ways\.  Once
you start with a word character that isn't a low line or a digit \(but other
\\p\{Pc\} and \\pN code points are ok though\, like U\+203F UNDERTIE and the Roman
numerals\)\, you can have anything you want \-\- \*if\* it's over Latin1\!\!  I've
tracked this down to a comment in toke\.c\, in the charname branch for source in
utf8\, which reads&#8203;:

&nbsp;     \* \.\.\. We accept anything above the latin1
&nbsp;     \* range because it is immaterial to Perl if it is
&nbsp;     \* correct or not\, and is expensive to check\.

So apparently this is intentional\.  Is this intentionally un\(der\)documented?

In any event\, I'm surprised you can start with underties or Roman numerals\. 
Yes\, the Roman numerals are alphabetic\, but the undertie seems strange\.
I think we're using word characters but claiming alphanums\.  

Actually\, it seems like you can use any >Latin1 char you wnat&#8203;:

&nbsp;    % blead \-Mutf8 \-le 'print ord "\\N\{⁶\}"'
&nbsp;    Unknown charname 'â¶' at \-e line 1\.
&nbsp;    65533

&nbsp;    % blead \-Mutf8 \-le 'print ord "\\N\{€\}"'
&nbsp;    Unknown charname 'â¬' at \-e line 1\.
&nbsp;    65533

&nbsp;    % blead \-Mutf8 \-le 'print ord "\\N\{👽\}"'
&nbsp;    Unknown charname 'ð½' at \-e line 1\.
&nbsp;    65533

Surely the stderr double\-encoding should go\!  It's because of STDERR already
being in UTF8\, because I have PERL\_UNICODE=SA\.  Here's the explicit override&#8203;:

&nbsp;    % blead \-C0 \-Mutf8 \-le 'print ord "\\N\{👽\}"'
&nbsp;    Unknown charname '👽' at \-e line 1\.
&nbsp;    65533
&nbsp;    % blead \-CS \-Mutf8 \-le 'print ord "\\N\{👽\}"'
&nbsp;    Unknown charname 'ð½' at \-e line 1\.
&nbsp;    65533

That's not right\.

Hm\, can I \*use\* a space alien for a name??

&nbsp;    % blead \-C0 \-Mutf8 \-wle 'use charnames "&#8203;:full"\, "&#8203;:alias" => \{ "👽" => "DOLLAR SIGN" \}; print ord "\\N\{👽\}"'
&nbsp;    Unknown charname '👽' at \-e line 1\.
&nbsp;    65533

Apparently not\.   But why didn't it warn me on the use charnames if didn't
like the "👽"  as a key?

Enough of that\.  Moving right along\.\.\.

Hm\, why is a bogus property a syntax error\, but a bogus charname isn't?

&nbsp;    % blead \-Mcharnames=&#8203;:full \-le 'print "\." =~ /\\p\{none such\}/ || "FAIL"'
&nbsp;    Can't find Unicode property definition "none such" at \-e line 1\.
&nbsp;    Exit 255

And why is this so inconsistent?

&nbsp;    % blead \-le 'print "\." =~ /\\x\{DEADBEEF CAKE\}/ || "FAIL"'
&nbsp;    FAIL

&nbsp;    % blead \-le 'print "\." =~ /\\N\{U\+DEADBEEF CAKE\}/ || "FAIL"'
&nbsp;    Invalid hexadecimal number in \\N\{U\+\.\.\.\} at \-e line 1\, within pattern
&nbsp;    Execution of \-e aborted due to compilation errors\.
&nbsp;    Exit 255

Surely the first should produce the same result as the second&#8203;: death\.

This is also inconsistent&#8203;:

&nbsp;    % perl \-Mdiagnostics \-wle 'print "\." =~ /\\x\{DEADBEEF CAKE\}/ || "FAIL"'
&nbsp;    Illegal hexadecimal digit ' ' ignored at \-e line 1 \(\#1\)
&nbsp;    \(W digit\) You may have tried to use a character other than 0 \- 9 or
&nbsp;    A \- F\, a \- f in a hexadecimal number\.  Interpretation of the hexadecimal
&nbsp;    number stopped before the illegal character\.
&nbsp;    
&nbsp;    FAIL

&nbsp;    % perl \-Mdiagnostics \-wle 'print "\." =~ /\\o\{DEADBEEF CAKE\}/ || "FAIL"'
&nbsp;    Non\-octal character 'D'\.  Resolved as "\\o\{\}" at \-e line 1 \(\#1\)
&nbsp;    \(W digit\)  In parsing an octal numeric constant\, a character was
&nbsp;    unexpectedly encountered that isn't octal\.  The resulting value is as
&nbsp;    indicated\.
&nbsp;    
&nbsp;    FAIL

And what's this about?

&nbsp;    % blead \-wle 'print 0x \- 1'
&nbsp;    \-1

&nbsp;    % blead \-wle 'print 0b \- 1'
&nbsp;    \-1

&nbsp;    % blead \-wle 'print 0xx2'
&nbsp;    00

Surely that's again an error?  Those are the same kind of thing as this&#8203;:

&nbsp;    % blead \-wle 'print 3\.14e'
&nbsp;    Bareword found where operator expected at \-e line 1\, near "3\.14e"
&nbsp;        \(Missing operator before e?\)
&nbsp;    Unquoted string "e" may clash with future reserved word at \-e line 1\.
&nbsp;    syntax error at \-e line 1\, near "3\.14e
&nbsp;    "
&nbsp;    Execution of \-e aborted due to compilation errors\.
&nbsp;    Exit 255

\-\-tom

\-\- 

Summary of my perl5 \(revision 5 version 14 subversion 0\) configuration&#8203;:
&nbsp;   
&nbsp;  Platform&#8203;:
&nbsp;    osname=openbsd\, osvers=4\.4\, archname=OpenBSD\.i386\-openbsd
&nbsp;    uname='openbsd chthon 4\.4 generic\#0 i386 '
&nbsp;    config\_args='\-des'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=undef\, usemultiplicity=undef
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=undef\, use64bitall=undef\, uselongdouble=undef
&nbsp;    usemymalloc=y\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'\,
&nbsp;    optimize='\-O2'\,
&nbsp;    cppflags='\-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='3\.3\.5 \(propolice\)'\, gccosandvers='openbsd4\.4'
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=4\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags ='\-Wl\,\-E  \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /usr/lib
&nbsp;    libs=\-lgdbm \-lm \-lutil \-lc
&nbsp;    perllibs=\-lm \-lutil \-lc
&nbsp;    libc=/usr/lib/libc\.so\.48\.0\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version=''
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags='\-DPIC \-fPIC '\, lddlflags='\-shared \-fPIC  \-L/usr/local/lib \-fstack\-protector'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;: MYMALLOC PERL\_DONT\_CREATE\_GVSV PERL\_MALLOC\_WRAP
&nbsp;                        PERL\_PRESERVE\_IVUV USE\_LARGE\_FILES USE\_PERLIO
&nbsp;                        USE\_PERL\_ATOF
&nbsp;  Built under openbsd
&nbsp;  Compiled at Jun 11 2011 11&#8203;:48&#8203;:28
&nbsp;  %ENV&#8203;:
&nbsp;    PERL\_UNICODE="SA"
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.14\.0/OpenBSD\.i386\-openbsd
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.14\.0
&nbsp;    /usr/local/lib/perl5/5\.14\.0/OpenBSD\.i386\-openbsd
&nbsp;    /usr/local/lib/perl5/5\.14\.0
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.12\.3
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.11\.3
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.10\.1
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.10\.0
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.8\.7
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.8\.0
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.6\.0
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.005
&nbsp;    /usr/local/lib/perl5/site\_perl
&nbsp;    \.

**Comment by p5pRT** [comment]
### From @khwilliamson

On 03/28/2012 02&#8203;:54 PM\, Tom Christiansen wrote&#8203;:
> SUMMARY&#8203;: Errors in \\x\{\.\.\.\} and \\N\{\.\.\.\} are handled poorly\, leading
>      to mysteriously inconsistent and incorrect results\.  Also\,
>           their are parsing bugs in \\N\{\.\.\.\.\} that permit all kinds of
>           stuff that it shouldn't\.
>

I'm the one responsible for many of the inconsistencies between things\. 
&nbsp;  I added parsing code for some new constructs\, like \\o\{\}\, and I did not 
want to repeat some of the mistakes I thought were made in\, say in \\x\{\}\. 
&nbsp;  IIRC\, no one complained about the exclusion of empty braces in the new 
constructs\, but when I proposed changing\, e\.g\. \\x\{\}\, to give an error or 
even a warning\, there were complaints that this would break existing 
code where the programmer likes that \{\} means 0\, or complicate their 
lives with noise from warnings that they don't want to get\.

If people didn't like that \\o\{\} requires at least one digit\, and had to 
be consistent with other constructs\, then it would be a simple matter to 
remove the prohibition; adding the prohibition later would run into the 
same backwards compat arguments\.

I think you would get more of a response to your initial email if you 
were to break it into smaller bits\, and/or write individual trouble tickets\.

> This seems wrong&#8203;:
>
>      $ blead \-wle 'print length "\\x\{\}"'
>      1
>      $ blead \-wle 'print ord "\\x\{\}"'
>      0
>
> The answer \*should\* be that that is altogether illegal\.  The stupid
> thing isn't even giving me a warning\.  What's that about?

The existing paradigm is that \{\} means 0\.  I don't like it\, but there 
was resistance to changing it\.  See
https://rt-archive.perl.org/perl5/Ticket/Display.html?id=72798

>
> Contrast with this much more reasonable behavior&#8203;:
>
>      $ blead \-wle 'print ord "\\o\{\}"'
>      Number with no digits at \-e line 1\, within string
>      Execution of \-e aborted due to compilation errors\.

>
> And I cannot see any justification for this&#8203;:
>
>      $ blead \-le 'print $n = \(\) = "8\\x\{00\}8" =~ /\\A\[\\8\]\{3\}\\z/'
>      1
>      $ blead \-le 'print $n = \(\) = "g\\x\{00\}g" =~ /\\A\[\\xg\]\{3\}\\z/'
>      1
>
> Yes\, those will generate optional ignorable warnings if you beg it nicely\,
> but they're \*syntax errors\*\.  It doesn't seem right to emit optional and
> by\-default\-off warnings on syntax errors\, for goodness' sake\!  The same
> thing should happen as with a bad \\o\{\} \-\- it should die\.

I agree\.  I thought there was a trouble ticket on it\, but couldn't find 
it\.  Anyway\, there was resistance to changing it\.
>
> This is a poor message\, for several reasons&#8203;:
>
>      % blead \-le 'print "\." =~ /\\N\{\}/ || "FAIL"'
>      Unknown charname '' at \-e line 1\.
>      Deprecated character in \\N\{\.\.\.\}; marked by\<\-\- HERE  in \\N\{\}\<\-\- HERE  at \-e line 1\.
>      FAIL
>
> First\, that looks a great deal like a syntax error to me\.  Why isn't
> it marked as such and the whole thing dying?  It's not just that it
> couldn't find a character name\.  It's that it wasn't even given one\.
> That should be any error\.

I agree\, and doubt that people would object to at least a better 
warning\.  Write a ticket\, and we'll see if there is dispute about it 
being an error\.
>
> There are other weirdness\.  I understand what's going on\, but it's
> a bit surprising\.
>
>      % blead \-le 'print "\." =~ /\\N\{$a\}/ || "FAIL"'
>      Unknown charname '$a' at \-e line 1\.
>      Deprecated character in \\N\{\.\.\.\}; marked by\<\-\- HERE  in \\N\{$\<\-\- HERE a\} at \-e line 1\.
>      FAIL
>
>      % blead \-le '$a = 1; print "\." =~ /\\N\{$a\}/ || "FAIL"'
>      Unknown charname '$a' at \-e line 1\.
>      Deprecated character in \\N\{\.\.\.\}; marked by\<\-\- HERE  in \\N\{$\<\-\- HERE a\} at \-e line 1\.
>      Name "main&#8203;::a" used only once&#8203;: possible typo at \-e line 1\.
>      FAIL
>
> Ok\, so variable interpolation doesn't happen in \\N\{brackets\}\.  Is that
> actually documented?

Include this in the ticket\.
>
> And what sort of warning \*is\* that?  How do you control it?
>
> This didn't turn it off&#8203;:
>
>      % blead \-X \-le 'print "\." =~ /\\N\{\}/ || "FAIL"'
>      Unknown charname '' at \-e line 1\.
>      FAIL
>
> Nor did this&#8203;:
>
>      % blead \-M\-warnings \-le 'print "\." =~ /\\N\{\}/ || "FAIL"'
>      Unknown charname '' at \-e line 1\.
>      FAIL
>
> And yet this fatalized it&#8203;:
>
>      % blead \-Mwarnings=FATAL\,all \-le 'print "\." =~ /\\N\{\}/ || "FAIL"'
>      Unknown charname '' at \-e line 1\.
>      Deprecated character in \\N\{\.\.\.\}; marked by\<\-\- HERE  in \\N\{\}\<\-\- HERE  at \-e line 1\.
>      Exit 255
>
> And this missed it altogether&#8203;:
>
>      % perl \-Mcharnames=&#8203;:full \-M\-warnings \-le 'print "\." =~ /\\N\{\}/ || "FAIL"'
>      Unknown charname '' at \-e line 1
>
>      FAIL
>
> So you can't turn it off in any of the normal ways\.
>
> It also seems to have an extra newline in the output\.
>
> Here diagnostics couldn't catch it&#8203;:
>
>      % blead \-Mcharnames=&#8203;:full \-Mdiagnostics \-le 'print "\." =~ /\\N\{\}/ || "FAIL"'
>      Unknown charname '' at \-e line 1
>
>      Deprecated character in \\N\{\.\.\.\}; marked by\<\-\- HERE  in \\N\{\}\<\-\- HERE  at \-e
>         line 1 \(\#1\)
>     \(D deprecated\) Just about anything is legal for the \.\.\. in \\N\{\.\.\.\}\.
>     But starting in 5\.12\, non\-reasonable ones that don't look like names
>     are deprecated\.  A reasonable name begins with an alphabetic character
>     and continues with any combination of alphanumerics\, dashes\, spaces\,
>     parentheses or colons\.
>     
>      FAIL

Include in ticket\.

>
> And there's all that extra newline business still\.
>
> That error message is misleading\.  It looks like it's saying
>
>      m\{
>     \\A \\p\{Alphabetic\}
>     \[\\p\{Alphabetic\}\\p\{Number\}\\p\{Dash\}\\N\{SPACE\}\(\)&#8203;:\] \+
>     \\z
>      \}z
>
>
> But that's incorrect\.  It appears to actually be this&#8203;:
>
>      m\{
>     \\A
>     \(?= \\p\{word\} \)            \# must start with a word character\, that isn't digit or \_
>     \(?\! \[\\N\{LOW LINE\}\\p\{digit\}\] \)   \# but \*does\* admit other numerics and connector punctuation\!\!
>     \[\\p\{word\}\\N\{HYPHEN\-MINUS\}\\N\{SPACE\}\(\)&#8203;:\] \*
>     \\z
>      \}x
>
> Which isn't what the message says\.  And that star really should be a
> plus\, as previously discussed\.

For non\-utf8 names\, the first character must be alphabetic\, and 
subsequent ones match qr/\[\-\\p\{XPosixWord\} \(\)&#8203;:\\xa0\]/\.

See below for utf8 name discussion\.

>
> Oh wait\.  It turns out that this is the real expressions&#8203;:
>
>      % setenv PERL\_UNICODE=SA
>
>      % perl \-Mcharnames=&#8203;:full \-le 'print qq\(print "\\\\\\N\{X\\N\{EURO SIGN\}\}"\)'
>      print "\\N\{X€\}"
>      % perl \-Mcharnames=&#8203;:full \-le 'print qq\(print "\\\\\\N\{X\\N\{EURO SIGN\}\}"\)' | blead \-Mutf8
>      Unknown charname 'Xâ¬' at \- line 1\.
>      ?
>      % perl \-le 'print qq\(print "\\\\\\N\{a\\x\{fffd\}\}"\)' | uniquote \-v
>      print "\\N\{a\\N\{REPLACEMENT CHARACTER\}\}"
>      % perl \-le 'print qq\(print "\\\\\\N\{a\\x\{fffd\}\}"\)' | uniquote \-x
>      print "\\N\{a\\x\{FFFD\}\}"
>      % perl \-le 'print qq\(print "\\\\\\N\{a\\x\{fffd\}\}"\)' | blead \-Mutf8
>      Unknown charname 'aï¿½' at \- line 1\.
>      % perl \-le 'print qq\(print "\\\\\\N\{a\\x\{fffd\}\#\#x%\}"\)' | blead \-Mutf8 \-l
>      Unknown charname 'aï¿½\#\#x%' at \- line 1\.
>      Deprecated character in \\N\{\.\.\.\}; marked by\<\-\- HERE  in \\N\{aï¿½\#\<\-\- HERE \#x%\} at \- line 1\.
>      ?
>      % perl \-Mcharnames=&#8203;:full \-le 'print qq\(print "\\\\\\N\{X\\N\{EURO SIGN\}\\N\{REPLACEMENT CHARACTER\}\\N\{PESO SIGN\}\\N\{ALIEN MONSTER\}\}"\)' | blead \-Mutf8 \-l
>      Unknown charname 'Xâ¬ï¿½â±ð¾' at \- line 1\.
>      ?
>      % perl \-Mcharnames=&#8203;:full \-le 'print qq\(print "\\\\\\N\{X\\N\{EURO SIGN\}\\N\{REPLACEMENT CHARACTER\}\\N\{PESO SIGN\}\\N\{ALIEN MONSTER\}\.\}"\)' | blead \-Mutf8 \-l
>      Unknown charname 'Xâ¬ï¿½â±ð¾\.' at \- line 1\.
>      Deprecated character in \\N\{\.\.\.\}; marked by\<\-\- HERE  in \\N\{Xâ¬ï¿½â±ð¾\.\<\-\- HERE \} at \-
>      line 1\.
>      ?
>
> First\, Perl is still double\-encoding UTF\-8 on stderr\.  Is that ever going
> to be fixed?
I'm not the one to fix it\.  Is there a ticket on this?

>
> But most importantly\, this is all messed up\, and in all kinds of ways\.  Once
> you start with a word character that isn't a low line or a digit \(but other
> \\p\{Pc\} and \\pN code points are ok though\, like U\+203F UNDERTIE and the Roman
> numerals\)\, you can have anything you want \-\- \*if\* it's over Latin1\!\!  I've
> tracked this down to a comment in toke\.c\, in the charname branch for source in
> utf8\, which reads&#8203;:
>
>       \* \.\.\. We accept anything above the latin1
>       \* range because it is immaterial to Perl if it is
>       \* correct or not\, and is expensive to check\.
>
> So apparently this is intentional\.  Is this intentionally un\(der\)documented?
>
> In any event\, I'm surprised you can start with underties or Roman numerals\.
> Yes\, the Roman numerals are alphabetic\, but the undertie seems strange\.
> I think we're using word characters but claiming alphanums\.
>
> Actually\, it seems like you can use any>Latin1 char you wnat&#8203;:
>
>      % blead \-Mutf8 \-le 'print ord "\\N\{⁶\}"'
>      Unknown charname 'â¶' at \-e line 1\.
>      65533
>
>      % blead \-Mutf8 \-le 'print ord "\\N\{€\}"'
>      Unknown charname 'â¬' at \-e line 1\.
>      65533
>
>      % blead \-Mutf8 \-le 'print ord "\\N\{👽\}"'
>      Unknown charname 'ð½' at \-e line 1\.
>      65533
>
> Surely the stderr double\-encoding should go\!  It's because of STDERR already
> being in UTF8\, because I have PERL\_UNICODE=SA\.  Here's the explicit override&#8203;:
>
>      % blead \-C0 \-Mutf8 \-le 'print ord "\\N\{👽\}"'
>      Unknown charname '👽' at \-e line 1\.
>      65533
>      % blead \-CS \-Mutf8 \-le 'print ord "\\N\{👽\}"'
>      Unknown charname 'ð½' at \-e line 1\.
>      65533
>
> That's not right\.
>
> Hm\, can I \*use\* a space alien for a name??
>
>      % blead \-C0 \-Mutf8 \-wle 'use charnames "&#8203;:full"\, "&#8203;:alias" =>  \{ "👽" =>  "DOLLAR SIGN" \}; print ord "\\N\{👽\}"'
>      Unknown charname '👽' at \-e line 1\.
>      65533
>
> Apparently not\.   But why didn't it warn me on the use charnames if didn't
> like the "👽"  as a key?

See https://rt-archive.perl.org/perl5/Ticket/Display.html?id=73022
This is to be fixed in 5\.18\.  We need to come up with the correct syntax 
for utf8 names\.  That is a whole discussion which I'd like to defer 
until 5\.16 is out the door\.  For example\, should it be xidstart and 
xidcontinue plus limited extensions?

Also see
https://rt-archive.perl.org/perl5/Ticket/Display.html?id=72712
which\, IIRC met some resistance on p5p which didn't make it into the ticket


>
> Enough of that\.  Moving right along\.\.\.
>
> Hm\, why is a bogus property a syntax error\, but a bogus charname isn't?
>
>      % blead \-Mcharnames=&#8203;:full \-le 'print "\." =~ /\\p\{none such\}/ || "FAIL"'
>      Can't find Unicode property definition "none such" at \-e line 1\.
>      Exit 255
>
> And why is this so inconsistent?
>
>      % blead \-le 'print "\." =~ /\\x\{DEADBEEF CAKE\}/ || "FAIL"'
>      FAIL
>
>      % blead \-le 'print "\." =~ /\\N\{U\+DEADBEEF CAKE\}/ || "FAIL"'
>      Invalid hexadecimal number in \\N\{U\+\.\.\.\} at \-e line 1\, within pattern
>      Execution of \-e aborted due to compilation errors\.
>      Exit 255
>
> Surely the first should produce the same result as the second&#8203;: death\.

I wrote the \\N\{U\+\.\.\.\} code to check for this\.  I think the first will 
give a warning if they are enabled\.  I disagree with the original design 
decision in \\x\{\} which is to just ignore anything after the first 
non\-hex digit before the brace\, and return the value computed so far\. 
This can't be the value that the programmer intended\, but Perl tries to 
soldier on\.  I claim you don't do anyone any favors by processing 
garbage as if it were real\, but that is what was done; and there may be 
resistance to fixing it\.
>
> This is also inconsistent&#8203;:
>
>      % perl \-Mdiagnostics \-wle 'print "\." =~ /\\x\{DEADBEEF CAKE\}/ || "FAIL"'
>      Illegal hexadecimal digit ' ' ignored at \-e line 1 \(\#1\)
>     \(W digit\) You may have tried to use a character other than 0 \- 9 or
>     A \- F\, a \- f in a hexadecimal number\.  Interpretation of the hexadecimal
>     number stopped before the illegal character\.
>
>      FAIL
>
>      % perl \-Mdiagnostics \-wle 'print "\." =~ /\\o\{DEADBEEF CAKE\}/ || "FAIL"'
>      Non\-octal character 'D'\.  Resolved as "\\o\{\}" at \-e line 1 \(\#1\)
>     \(W digit\)  In parsing an octal numeric constant\, a character was
>     unexpectedly encountered that isn't octal\.  The resulting value is as
>     indicated\.
>
>      FAIL

This looks to be my \\o\{\} fault; file a ticket

>
> And what's this about?
>
>      % blead \-wle 'print 0x \- 1'
>      \-1
>
>      % blead \-wle 'print 0b \- 1'
>      \-1
>
>      % blead \-wle 'print 0xx2'
>      00
>
> Surely that's again an error?  Those are the same kind of thing as this&#8203;:
>
>      % blead \-wle 'print 3\.14e'
>      Bareword found where operator expected at \-e line 1\, near "3\.14e"
>         \(Missing operator before e?\)
>      Unquoted string "e" may clash with future reserved word at \-e line 1\.
>      syntax error at \-e line 1\, near "3\.14e
>      "
>      Execution of \-e aborted due to compilation errors\.
>      Exit 255
>

File a ticket\, but there may be resistance\.

**Comment by p5pRT** [comment]
@jkeenan - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @khwilliamson

This ticket was opened by mistake; I don't know how it happened

\-\- 
Karl Williamson

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'open' to 'rejected'

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Apr 05 19&#8203;:57&#8203;:39 2012\, khw wrote&#8203;:
> This ticket was opened by mistake; I don't know how it happened

Tom Christiansen’s original bug report was CC’ed to you as well as being
sent to perlbug@&#8203;perl\.org\.  So your copy did not have the bug number\.  I
imagine you did ‘Reply all’\, which would have CC’ed your response to
perlbug \(without the bug number\)\, opening a new ticket\.  I merged the
two tickets together before you marked it as rejected\.  I’ll leave it to
you to decide whether to re\-open it\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Thu Apr 05 19&#8203;:57&#8203;:39 2012\, khw wrote&#8203;:
> This ticket was opened by mistake; I don't know how it happened

Tom Christiansen’s original bug report was CC’ed to you as well as being
sent to perlbug@&#8203;perl\.org\.  So your copy did not have the bug number\.  I
imagine you did ‘Reply all’\, which would have CC’ed your response to
perlbug \(without the bug number\)\, opening a new ticket\.  I merged the
two tickets together before you marked it as rejected\.  I’ll leave it to
you to decide whether to re\-open it\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @khwilliamson

This was wrongly rejected by a timing issue with two people working on RT around the same time\.  And it didn't get re\-opened until now\.  Some of the things have been fixed\, but some haven't\, and some won't be\.

I'll also 'take' this ticket to remind me to look at it

\-\- 
Karl Williamson

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'rejected' to 'open'

**Comment by p5pRT** [comment]
### From @khwilliamson

On Thu\, 05 Apr 2012 11&#8203;:03&#8203;:25 \-0700\, public@&#8203;khwilliamson\.com wrote&#8203;:
> On 03/28/2012 02&#8203;:54 PM\, Tom Christiansen wrote&#8203;:
> > SUMMARY&#8203;: Errors in \\x\{\.\.\.\} and \\N\{\.\.\.\} are handled poorly\, leading
> >        to mysteriously inconsistent and incorrect results\.  Also\,
> >           their are parsing bugs in \\N\{\.\.\.\.\} that permit all kinds of
> >           stuff that it shouldn't\.
> >
> 
> I'm the one responsible for many of the inconsistencies between
> things\.
>   I added parsing code for some new constructs\, like \\o\{\}\, and I did
> not
> want to repeat some of the mistakes I thought were made in\, say in
> \\x\{\}\.
>   IIRC\, no one complained about the exclusion of empty braces in the
> new
> constructs\, but when I proposed changing\, e\.g\. \\x\{\}\, to give an error
> or
> even a warning\, there were complaints that this would break existing
> code where the programmer likes that \{\} means 0\, or complicate their
> lives with noise from warnings that they don't want to get\.
> 
> If people didn't like that \\o\{\} requires at least one digit\, and had
> to
> be consistent with other constructs\, then it would be a simple matter
> to
> remove the prohibition; adding the prohibition later would run into
> the
> same backwards compat arguments\.
> 
> I think you would get more of a response to your initial email if you
> were to break it into smaller bits\, and/or write individual trouble
> tickets\.
> 
> > This seems wrong&#8203;:
> >
> > $ blead \-wle 'print length "\\x\{\}"'
> > 1
> > $ blead \-wle 'print ord "\\x\{\}"'
> > 0
> >
> > The answer \*should\* be that that is altogether illegal\.  The stupid
> > thing isn't even giving me a warning\.  What's that about?
> 
> The existing paradigm is that \{\} means 0\.  I don't like it\, but there
> was resistance to changing it\.  See
> https://rt-archive.perl.org/perl5/Ticket/Display.html?id=72798
> 
> >
> > Contrast with this much more reasonable behavior&#8203;:
> >
> > $ blead \-wle 'print ord "\\o\{\}"'
> > Number with no digits at \-e line 1\, within string
> > Execution of \-e aborted due to compilation errors\.
> 
> >
> > And I cannot see any justification for this&#8203;:
> >
> > $ blead \-le 'print $n = \(\) = "8\\x\{00\}8" =~ /\\A\[\\8\]\{3\}\\z/'
> > 1
> > $ blead \-le 'print $n = \(\) = "g\\x\{00\}g" =~ /\\A\[\\xg\]\{3\}\\z/'
> > 1
> >
> > Yes\, those will generate optional ignorable warnings if you beg it
> > nicely\,
> > but they're \*syntax errors\*\.  It doesn't seem right to emit optional
> > and
> > by\-default\-off warnings on syntax errors\, for goodness' sake\!  The
> > same
> > thing should happen as with a bad \\o\{\} \-\- it should die\.
> 
> I agree\.  I thought there was a trouble ticket on it\, but couldn't
> find
> it\.  Anyway\, there was resistance to changing it\.
> >
> > This is a poor message\, for several reasons&#8203;:
> >
> > % blead \-le 'print "\." =~ /\\N\{\}/ || "FAIL"'
> > Unknown charname '' at \-e line 1\.
> > Deprecated character in \\N\{\.\.\.\}; marked by\<\-\- HERE  in \\N\{\}\<\-\- HERE
> > at \-e line 1\.
> > FAIL
> >
> > First\, that looks a great deal like a syntax error to me\.  Why isn't
> > it marked as such and the whole thing dying?  It's not just that it
> > couldn't find a character name\.  It's that it wasn't even given one\.
> > That should be any error\.
> 
> I agree\, and doubt that people would object to at least a better
> warning\.  Write a ticket\, and we'll see if there is dispute about it
> being an error\.

The empty name is now a syntax error\, so now you get

 Unknown charname '' at \-e line 1\, within pattern
 Execution of \-e aborted due to compilation errors\.

I think this is good enough\.

> >
> > There are other weirdness\.  I understand what's going on\, but it's
> > a bit surprising\.
> >
> > % blead \-le 'print "\." =~ /\\N\{$a\}/ || "FAIL"'
> > Unknown charname '$a' at \-e line 1\.
> > Deprecated character in \\N\{\.\.\.\}; marked by\<\-\- HERE  in \\N\{$\<\-\- HERE
> > a\} at \-e line 1\.
> > FAIL
> >
> > % blead \-le '$a = 1; print "\." =~ /\\N\{$a\}/ || "FAIL"'
> > Unknown charname '$a' at \-e line 1\.
> > Deprecated character in \\N\{\.\.\.\}; marked by\<\-\- HERE  in \\N\{$\<\-\- HERE
> > a\} at \-e line 1\.
> > Name "main&#8203;::a" used only once&#8203;: possible typo at \-e line 1\.
> > FAIL
> >
> > Ok\, so variable interpolation doesn't happen in \\N\{brackets\}\.  Is
> > that
> > actually documented?

Sort of\.  I will add a patch to more explicitly say so\.  Again\, these are now fatal errors\, so you get

 Unknown charname '$a' at \-e line 1\, within pattern
 Execution of \-e aborted due to compilation errors\.

> 
> Include this in the ticket\.
> >
> > And what sort of warning \*is\* that?  How do you control it?

It's now a fatal error

> >
> > This didn't turn it off&#8203;:
> >
> > % blead \-X \-le 'print "\." =~ /\\N\{\}/ || "FAIL"'
> > Unknown charname '' at \-e line 1\.
> > FAIL
> >
> > Nor did this&#8203;:
> >
> > % blead \-M\-warnings \-le 'print "\." =~ /\\N\{\}/ || "FAIL"'
> > Unknown charname '' at \-e line 1\.
> > FAIL
> >
> > And yet this fatalized it&#8203;:
> >
> > % blead \-Mwarnings=FATAL\,all \-le 'print "\." =~ /\\N\{\}/ || "FAIL"'
> > Unknown charname '' at \-e line 1\.
> > Deprecated character in \\N\{\.\.\.\}; marked by\<\-\- HERE  in \\N\{\}\<\-\- HERE
> > at \-e line 1\.
> > Exit 255
> >
> > And this missed it altogether&#8203;:
> >
> > % perl \-Mcharnames=&#8203;:full \-M\-warnings \-le 'print "\." =~ /\\N\{\}/ ||
> > "FAIL"'
> > Unknown charname '' at \-e line 1
> >
> > FAIL
> >
> > So you can't turn it off in any of the normal ways\.
> >
> > It also seems to have an extra newline in the output\.
> >
> > Here diagnostics couldn't catch it&#8203;:
> >
> > % blead \-Mcharnames=&#8203;:full \-Mdiagnostics \-le 'print "\." =~ /\\N\{\}/ ||
> > "FAIL"'
> > Unknown charname '' at \-e line 1
> >
> > Deprecated character in \\N\{\.\.\.\}; marked by\<\-\- HERE  in \\N\{\}\<\-\- HERE
> > at \-e
> >      line 1 \(\#1\)
> >  \(D deprecated\) Just about anything is legal for the \.\.\. in \\N\{\.\.\.\}\.
> >  But starting in 5\.12\, non\-reasonable ones that don't look like names
> >  are deprecated\.  A reasonable name begins with an alphabetic
> > character
> >  and continues with any combination of alphanumerics\, dashes\, spaces\,
> >  parentheses or colons\.
> >
> > FAIL
> 
> Include in ticket\.
> 
> >
> > And there's all that extra newline business still\.
> >
> > That error message is misleading\.  It looks like it's saying
> >
> > m\{
> >  \\A \\p\{Alphabetic\}
> >  \[\\p\{Alphabetic\}\\p\{Number\}\\p\{Dash\}\\N\{SPACE\}\(\)&#8203;:\] \+
> >  \\z
> > \}z
> >
> >
> > But that's incorrect\.  It appears to actually be this&#8203;:
> >
> > m\{
> >  \\A
> >  \(?= \\p\{word\} \)                  \# must start with a word character\,
> > that isn't digit or \_
> >  \(?\! \[\\N\{LOW LINE\}\\p\{digit\}\] \)   \# but \*does\* admit other numerics
> > and connector punctuation\!\!
> >  \[\\p\{word\}\\N\{HYPHEN\-MINUS\}\\N\{SPACE\}\(\)&#8203;:\] \*
> >  \\z
> > \}x
> >
> > Which isn't what the message says\.  And that star really should be a
> > plus\, as previously discussed\.

Again\, this has all been fixed by making \\N\{\} fatal\.
> 
> For non\-utf8 names\, the first character must be alphabetic\, and
> subsequent ones match qr/\[\-\\p\{XPosixWord\} \(\)&#8203;:\\xa0\]/\.
> 
> See below for utf8 name discussion\.
> 
> >
> > Oh wait\.  It turns out that this is the real expressions&#8203;:
> >
> > % setenv PERL\_UNICODE=SA
> >
> > % perl \-Mcharnames=&#8203;:full \-le 'print qq\(print "\\\\\\N\{X\\N\{EURO SIGN\}\}"\)'
> > print "\\N\{X€\}"
> > % perl \-Mcharnames=&#8203;:full \-le 'print qq\(print "\\\\\\N\{X\\N\{EURO SIGN\}\}"\)'
> > | blead \-Mutf8
> > Unknown charname 'Xâ¬' at \- line 1\.
> > ?
> > % perl \-le 'print qq\(print "\\\\\\N\{a\\x\{fffd\}\}"\)' | uniquote \-v
> > print "\\N\{a\\N\{REPLACEMENT CHARACTER\}\}"
> > % perl \-le 'print qq\(print "\\\\\\N\{a\\x\{fffd\}\}"\)' | uniquote \-x
> > print "\\N\{a\\x\{FFFD\}\}"
> > % perl \-le 'print qq\(print "\\\\\\N\{a\\x\{fffd\}\}"\)' | blead \-Mutf8
> > Unknown charname 'aï¿½' at \- line 1\.
> > % perl \-le 'print qq\(print "\\\\\\N\{a\\x\{fffd\}\#\#x%\}"\)' | blead \-Mutf8 \-l
> > Unknown charname 'aï¿½\#\#x%' at \- line 1\.
> > Deprecated character in \\N\{\.\.\.\}; marked by\<\-\- HERE  in \\N\{aï¿½\#\<\-\-
> > HERE \#x%\} at \- line 1\.
> > ?
> > % perl \-Mcharnames=&#8203;:full \-le 'print qq\(print "\\\\\\N\{X\\N\{EURO
> > SIGN\}\\N\{REPLACEMENT CHARACTER\}\\N\{PESO SIGN\}\\N\{ALIEN MONSTER\}\}"\)' |
> > blead \-Mutf8 \-l
> > Unknown charname 'Xâ¬ï¿½â±ð¾' at \- line 1\.
> > ?
> > % perl \-Mcharnames=&#8203;:full \-le 'print qq\(print "\\\\\\N\{X\\N\{EURO
> > SIGN\}\\N\{REPLACEMENT CHARACTER\}\\N\{PESO SIGN\}\\N\{ALIEN MONSTER\}\.\}"\)' |
> > blead \-Mutf8 \-l
> > Unknown charname 'Xâ¬ï¿½â±ð¾\.' at \- line 1\.
> > Deprecated character in \\N\{\.\.\.\}; marked by\<\-\- HERE  in
> > \\N\{Xâ¬ï¿½â±ð¾\.\<\-\- HERE \} at \-
> > line 1\.
> > ?
> >
> > First\, Perl is still double\-encoding UTF\-8 on stderr\.  Is that ever
> > going
> > to be fixed?
> I'm not the one to fix it\.  Is there a ticket on this?

I just tried these\, and the double encoding is gone\.
> 
> >
> > But most importantly\, this is all messed up\, and in all kinds of
> > ways\.  Once
> > you start with a word character that isn't a low line or a digit \(but
> > other
> > \\p\{Pc\} and \\pN code points are ok though\, like U\+203F UNDERTIE and
> > the Roman
> > numerals\)\, you can have anything you want \-\- \*if\* it's over Latin1\!\!
> > I've
> > tracked this down to a comment in toke\.c\, in the charname branch for
> > source in
> > utf8\, which reads&#8203;:
> >
> > \* \.\.\. We accept anything above the latin1
> > \* range because it is immaterial to Perl if it is
> > \* correct or not\, and is expensive to check\.
> >
> > So apparently this is intentional\.  Is this intentionally
> > un\(der\)documented?

This has all been fixed\.  The checking is done\.
> >
> > In any event\, I'm surprised you can start with underties or Roman
> > numerals\.
> > Yes\, the Roman numerals are alphabetic\, but the undertie seems
> > strange\.
> > I think we're using word characters but claiming alphanums\.

This has been fixed

> >
> > Actually\, it seems like you can use any>Latin1 char you wnat&#8203;:
> >
> > % blead \-Mutf8 \-le 'print ord "\\N\{⁶\}"'
> > Unknown charname 'â¶' at \-e line 1\.
> > 65533
> >
> > % blead \-Mutf8 \-le 'print ord "\\N\{€\}"'
> > Unknown charname 'â¬' at \-e line 1\.
> > 65533
> >
> > % blead \-Mutf8 \-le 'print ord "\\N\{👽\}"'
> > Unknown charname 'ð½' at \-e line 1\.
> > 65533
> >
> > Surely the stderr double\-encoding should go\!  It's because of STDERR
> > already
> > being in UTF8\, because I have PERL\_UNICODE=SA\.  Here's the explicit
> > override&#8203;:
> >
> > % blead \-C0 \-Mutf8 \-le 'print ord "\\N\{👽\}"'
> > Unknown charname '👽' at \-e line 1\.
> > 65533
> > % blead \-CS \-Mutf8 \-le 'print ord "\\N\{👽\}"'
> > Unknown charname 'ð½' at \-e line 1\.
> > 65533
> >
> > That's not right\.
> >
> > Hm\, can I \*use\* a space alien for a name??
> >
> > % blead \-C0 \-Mutf8 \-wle 'use charnames "&#8203;:full"\, "&#8203;:alias" =>  \{ "👽" =>
> > "DOLLAR SIGN" \}; print ord "\\N\{👽\}"'
> > Unknown charname '👽' at \-e line 1\.
> > 65533
> >
> > Apparently not\.   But why didn't it warn me on the use charnames if
> > didn't
> > like the "👽"  as a key?

All fixed\.

> 
> See https://rt-archive.perl.org/perl5/Ticket/Display.html?id=73022
>  This is to be fixed in 5\.18\.  We need to come up with the correct
> syntax
>  for utf8 names\.  That is a whole discussion which I'd like to defer
>  until 5\.16 is out the door\.  For example\, should it be xidstart and
> xidcontinue plus limited extensions?
> 
> Also see
> https://rt-archive.perl.org/perl5/Ticket/Display.html?id=72712
> which\, IIRC met some resistance on p5p which didn't make it into the
> ticket
> 
> 
> >
> > Enough of that\.  Moving right along\.\.\.
> >
> > Hm\, why is a bogus property a syntax error\, but a bogus charname
> > isn't?
> >
> > % blead \-Mcharnames=&#8203;:full \-le 'print "\." =~ /\\p\{none such\}/ ||
> > "FAIL"'
> > Can't find Unicode property definition "none such" at \-e line 1\.
> > Exit 255

Now both are syntax errors
> >
> > And why is this so inconsistent?
> >
> > % blead \-le 'print "\." =~ /\\x\{DEADBEEF CAKE\}/ || "FAIL"'
> > FAIL

This is now a syntax error
> >
> > % blead \-le 'print "\." =~ /\\N\{U\+DEADBEEF CAKE\}/ || "FAIL"'
> > Invalid hexadecimal number in \\N\{U\+\.\.\.\} at \-e line 1\, within pattern
> > Execution of \-e aborted due to compilation errors\.
> > Exit 255
> >
> > Surely the first should produce the same result as the second&#8203;: death\.

Now they both do\.
> 
> I wrote the \\N\{U\+\.\.\.\} code to check for this\.  I think the first will
> give a warning if they are enabled\.  I disagree with the original
> design
> decision in \\x\{\} which is to just ignore anything after the first
> non\-hex digit before the brace\, and return the value computed so far\.
> This can't be the value that the programmer intended\, but Perl tries
> to
> soldier on\.  I claim you don't do anyone any favors by processing
> garbage as if it were real\, but that is what was done; and there may
> be
> resistance to fixing it\.
> >
> > This is also inconsistent&#8203;:
> >
> > % perl \-Mdiagnostics \-wle 'print "\." =~ /\\x\{DEADBEEF CAKE\}/ ||
> > "FAIL"'
> > Illegal hexadecimal digit ' ' ignored at \-e line 1 \(\#1\)
> >  \(W digit\) You may have tried to use a character other than 0 \- 9 or
> >  A \- F\, a \- f in a hexadecimal number\.  Interpretation of the
> > hexadecimal
> >  number stopped before the illegal character\.
> >
> > FAIL
> >
> > % perl \-Mdiagnostics \-wle 'print "\." =~ /\\o\{DEADBEEF CAKE\}/ ||
> > "FAIL"'
> > Non\-octal character 'D'\.  Resolved as "\\o\{\}" at \-e line 1 \(\#1\)
> >  \(W digit\)  In parsing an octal numeric constant\, a character was
> >  unexpectedly encountered that isn't octal\.  The resulting value is
> > as
> >  indicated\.
> >
> > FAIL
> 
> This looks to be my \\o\{\} fault; file a ticket

This is still the case\.  I think the current behavior is acceptable\, given the warnings generated\.  If you want it changed\, file a ticket
> 
> >
> > And what's this about?
> >
> > % blead \-wle 'print 0x \- 1'
> > \-1
> >
> > % blead \-wle 'print 0b \- 1'
> > \-1
> >
> > % blead \-wle 'print 0xx2'
> > 00
> >
> > Surely that's again an error?  Those are the same kind of thing as
> > this&#8203;:
> >
> > % blead \-wle 'print 3\.14e'
> > Bareword found where operator expected at \-e line 1\, near "3\.14e"
> >      \(Missing operator before e?\)
> > Unquoted string "e" may clash with future reserved word at \-e line 1\.
> > syntax error at \-e line 1\, near "3\.14e
> > "
> > Execution of \-e aborted due to compilation errors\.
> > Exit 255
> >
> 
> File a ticket\, but there may be resistance\.


I filed https://rt-archive.perl.org/perl5/Ticket/Display.html?id=134125

\-\- 
Karl Williamson

**Comment by p5pRT** [comment]
### From @khwilliamson

I am resolving this ticket\, which lists many different issues\.  All are fixed\, rejected\, or in another ticket
\-\- 
Karl Williamson

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'open' to 'resolved'
